src/Conversion/ONNXToKrnl/Math/Reduction.cpp

```diff
@@ -150,16 +151,40 @@ struct ONNXReductionOpLowering : public ConversionPattern {
     int64_t inRank = memRefInType.getRank();
     int64_t outRank = memRefOutType.getRank();
 
-    // Get attributes
-    ArrayAttr axisAttrs = llvm::dyn_cast<ONNXReductionOp>(op).axesAttr();
+    // Get axes value defined by op
+    // Leave empty is not defined
+    std::vector<int64_t> definedAxes;
+    if constexpr (std::is_same<ONNXReductionOp, ONNXReduceSumOp>::value) {
+      // Assume it is verified that axes are known
+      // Convert DenseElementsAttr to ArrayAttr
+      Value axesValue = llvm::dyn_cast<ONNXReductionOp>(op).axes();
+      if (getONNXConstantOp(axesValue)) {
+        DenseElementsAttr constAxes =
+            getONNXConstantOp(axesValue)
+                .valueAttr()
+                .dyn_cast_or_null<mlir::DenseElementsAttr>();
+        SmallVector<int64_t, 4> values;
+        for (auto element : constAxes.getValues<IntegerAttr>()) {
+          definedAxes.push_back(element.getInt());
+        }
+      }
+    } else {
+      ArrayAttr axisAttrs = llvm::dyn_cast<ONNXReductionOp>(op).axesAttr();
+      if (axisAttrs) {
+        for (auto axisAttr : axisAttrs.getValue()) {
+          int64_t axis = axisAttr.cast<IntegerAttr>().getInt();
+          definedAxes.push_back(axis);
+        }
+      }
+    }
+
     std::vector<int64_t> axes;
-    if (axisAttrs) {
-      for (auto axisAttr : axisAttrs.getValue()) {
-        int64_t axis = axisAttr.cast<IntegerAttr>().getInt();
-        axis = axis >= 0 ? axis : (inRank + axis);
-        assert(axis >= -inRank && axis <= inRank - 1);
-        if (std::find(axes.begin(), axes.end(), axis) == axes.end())
-          axes.push_back(axis);
+    if (definedAxes.size()) {
+      for (auto axis : definedAxes) {
+        int64_t newaxis = axis >= 0 ? axis : (inRank + axis);
+        assert(newaxis >= -inRank && newaxis <= inRank - 1);
```

Comment:
When the inputs are bad, don't we generally emit an error instead of an assert? I though that was our policy. Since this come from the "user".

Reply:
This is the case for optional input. The axes input may be missing or empty. We have to handle it.

Reply:
First, newaxis>=0 && <= inRank-1  so the check must be updated. Or the current test should be tested on "axis" and not "newaxis"

Second, you have an assert here. What I am saying is that our policy is to use assert for compiler errors, and emitError for model errors. When checking values that come from the model, we should thus use emitError.

Reply:
fixed.